### PR TITLE
[Torchvision API] resized_crop and RandomResizedCrop

### PR DESCRIPTION
**Pull request by mdabek-nvidia**
## Category:

**New feature**

## Description:

Torchvision API operators: 

* RandomResizedCrop
* Functional resized_crop functional
* Unification of InterpolationMode validation

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [X] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [X] Docstring
  - [ ] Doxygen
  - [X] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by mdabek-nvidia** [comment]
@greptileai review please

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    left: int,
+    height: int,
+    width: int,
+    device: DeviceLike = "cpu",
+) -> ndd.Tensor | ndd.Batch:
+    """
+    Please refer to the ``RandomCrop`` operator for more details.
+    """
+    _verify_crop_args(top, left, height, width)
+
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:108]: This is inline with original Torchvision comment

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    Verify RandomResizedCrop interpolation argument.
+    """
+
+    @classmethod
+    def verify(cls, *, interpolation, **_) -> None:
+        if interpolation in Resize.not_supported_interpolation_modes:
+            raise NotImplementedError(f"Interpolation mode: {interpolation} is not supported")
+        if interpolation not in Resize.interpolation_modes:
+            raise ValueError(f"Interpolation {type(interpolation)} is not supported")
+
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:121]: Fixed

File: dali/test/python/torchvision/test_tv_randomresizedcrop.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+from nose2.tools import cartesian_params, params
+from nose_utils import assert_raises
+import numpy as np
+from PIL import Image
+import torch
+import torchvision.transforms.v2 as transforms
+import torchvision.transforms.v2.functional as tv_fn
+
+import nvidia.dali.experimental.torchvision.v2.randomcrop as randomcrop_module
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/test/python/torchvision/test_tv_randomresizedcrop.py:26]: This is test code

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+def _verify_crop_args(top: int, left: int, height: int, width: int) -> None:
+    _verify_crop_coordinate(top, "top")
+    _verify_crop_coordinate(left, "left")
+    RandomCrop.verify_args(
+        size=(height, width),
+        padding=None,
+        pad_if_needed=False,
+        padding_mode="constant",
+        fill=0,
+    )
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py]: Fixed

**Comment by mdabek-nvidia** [comment]
@greptileai re-review

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 5]
```diff
@@ -13,13 +13,22 @@
...
+
 from nvidia.dali._typing import TensorLike
 from nvidia.dali.experimental.dynamic._device import DeviceLike
 
 from ..operator import adjust_input
 from ..randomcrop import RandomCrop
+from ..resize import Resize
+
+
+def _verify_crop_coordinate(value, name: str) -> None:
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py]: **[Minor] Dead code.** `_verify_crop_coordinate` is defined here but never called — `_validate_crop_params` uses `_validate_integer_param` for the coordinate checks (and the tests assert the `"must be an integer"` message it produces, not the `"must be int"` message here). Looks like a leftover from refactoring; please remove it.

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 6]
```diff
@@ -86,6 +88,41 @@ def verify(cls, *, fill, **_) -> None:
...
+        if value[0] > value[1]:
+            raise ValueError(f"{name} should be a (min, max) range, got {value}")
+
+    @classmethod
+    def verify(cls, *, scale, ratio, **_) -> None:
+        cls._verify_range(scale, "scale")
+        cls._verify_range(ratio, "ratio")
+
+
+class _ValidateRandomResizedCropInterpolation(_ArgumentValidateRule):
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:113]: **[Nit / DRY]** This interpolation check is now the third copy of the same logic: it duplicates `Resize._ValidateSize` (resize.py lines 57-61) and partly overlaps with `Resize.normalize_interpolation`. The three paths can silently drift apart if the supported/not-supported sets change. Consider exposing a single classmethod on `Resize` (e.g. `Resize.verify_interpolation(interpolation)`) and calling it from all three sites. Not blocking.

**Review by jantonguirao — APPROVED** [review]

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py (hunk 6 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:113]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py (hunk 5 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/resize.py [hunk 7]
```diff
@@ -98,9 +104,39 @@ class Resize(Operator):
...
+
     arg_rules = [_ValidateSize]
     preprocess_data = get_HWC_from_layout_pipeline
 
+    @classmethod
+    def validate_interpoliation(cls, interpolation) -> None:
+        if interpolation in cls.not_supported_interpolation_modes:
+            raise NotImplementedError(f"Interpolation mode: {interpolation} is not supported")
+        if interpolation not in cls.interpolation_modes:
+            raise ValueError(f"Interpolation {interpolation!r} is not supported")
```
Comment (rostan-t) [inline dali/python/nvidia/dali/experimental/torchvision/v2/resize.py:126]: Nitpick: we force `__repr__` while the exception above doesn't?

File: dali/python/nvidia/dali/experimental/torchvision/v2/resize.py [hunk 8]
```diff
@@ -98,9 +104,39 @@ class Resize(Operator):
         InterpolationMode.HAMMING,
     ]
 
+    # Legacy PIL integer codes accepted by torchvision for back-compat
+    # (mirrors torchvision.transforms.functional._interpolation_modes_from_int).
```
Comment (rostan-t) [inline dali/python/nvidia/dali/experimental/torchvision/v2/resize.py:108]: Shouldn't we use `torchvision.transforms.functional._interpolation_modes_from_int` directly?

File: dali/test/python/torchvision/test_tv_crop.py [hunk 9]
```diff
@@ -130,9 +130,9 @@ def test_crop_preserves_tensor_dtype(dtype):
     dict(top=0, left=0, height=1, width=1.0),
 )
 def test_crop_tensor_rejects_float_parameters(crop_kwargs):
-    with assert_raises(TypeError):
+    with assert_raises(TypeError, glob="*integer*"):
```
Comment (rostan-t) [inline dali/test/python/torchvision/test_tv_crop.py]: Nitpick: surrounding with wildcards shouldn't be necessary.

File: dali/test/python/torchvision/test_tv_randomresizedcrop.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    assert out.shape == tv_out.shape, f"Shape mismatch: {out.shape} != {tv_out.shape}"
+    assert torch.allclose(out, tv_out, rtol=0, atol=atol), "DALI output differs from torchvision"
+
+
+def test_random_resized_crop_is_operator():
+    assert issubclass(RandomResizedCrop, Operator)
+
+
+def test_random_resized_crop_exported_from_randomcrop_module():
+    from nvidia.dali.experimental.torchvision.v2.randomcrop import RandomResizedCrop as exported
```
Comment (rostan-t) [inline dali/test/python/torchvision/test_tv_randomresizedcrop.py]: Why does it have to be exported from `v2.randomcrop`? I don't see this in TorchVision.

File: dali/test/python/torchvision/test_tv_randomresizedcrop.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+    out = _to_tensor(out)
+    tv_out = _to_tensor(tv_out)
+    if device == "gpu":
+        out = out.cpu()
+        if isinstance(tv_out, torch.Tensor):
+            tv_out = tv_out.cpu()
+
+    assert out.shape == tv_out.shape, f"Shape mismatch: {out.shape} != {tv_out.shape}"
+    assert torch.allclose(out, tv_out, rtol=0, atol=atol), "DALI output differs from torchvision"
```
Comment (rostan-t) [inline dali/test/python/torchvision/test_tv_randomresizedcrop.py]: Nitpick: `torch.testing.assert_close` provides more information in case of mismatch.

Comment (mdabek-nvidia) [reply dali/test/python/torchvision/test_tv_randomresizedcrop.py]: Fixed

File: dali/test/python/torchvision/test_tv_randomresizedcrop.py (hunk 10 above)
Comment (mdabek-nvidia) [reply dali/test/python/torchvision/test_tv_randomresizedcrop.py]: I will remove this test, it does not seem to have much sense.

File: dali/test/python/torchvision/test_tv_crop.py (hunk 9 above)
Comment (mdabek-nvidia) [reply dali/test/python/torchvision/test_tv_crop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/resize.py (hunk 8 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/resize.py:108]: I want to move from Torchvision dependency towards DALI being one line replacement. I am yet to introduce interpolations mode.

File: dali/python/nvidia/dali/experimental/torchvision/v2/resize.py (hunk 7 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/resize.py:126]: Fixed.

**Comment by mdabek-nvidia** [comment]
!build

**Review by rostan-t — APPROVED** [review]
